### PR TITLE
Trace Logging on Permission Denied & ColorFormat

### PR DESCRIPTION
**Pull request by zeripath**
This PR adds `log.Trace(...)` calls to Permission Denied events.

It also adds a new additional way for providing colored logs for structs in a semi-standardised way. e.g.

```go
log.Info("Log me a user %-v", user)
```
Will log `Log me a user userID:username` with the userID being colored cyan and the username in bold.

Structs that want to implement this need to provide a `ColorFormat(s fmt.State)` function which is very similar to the `Format(...)` function expected by `fmt` (mainly lacking the rune - which would always be `'v'`.) If you logged `%-v` then this function will be called, and if more verbose output is possible a `+` should be passed in the flags and inspected for with `s.Flags('+')` in `ColorFormat`, e.g. `log.Info("%-+v", perm)` to log a verbose version of `models.Permission`.

### CODE REVIEW

File: models/repo_permission.go [hunk 1]
```diff
@@ -90,12 +96,69 @@ func (p *Permission) CanWriteIssuesOrPulls(isPull bool) bool {
...
+
 // GetUserRepoPermission returns the user permissions to the repository
 func GetUserRepoPermission(repo *Repository, user *User) (Permission, error) {
 	return getUserRepoPermission(x, repo, user)
 }
 
 func getUserRepoPermission(e Engine, repo *Repository, user *User) (perm Permission, err error) {
+	if log.IsTrace() {
+		defer func() {
+			if log.IsTrace() {
```
Comment (techknowlogick) [inline models/repo_permission.go]: Are two `if log.IsTrace() {` statements needed?

Comment (zeripath) [reply models/repo_permission.go]: I guess if we decide in future to make logs reloadable then we'll skip that call.

I can remove it if wanted.

File: modules/context/permission.go [hunk 2]
```diff
@@ -6,6 +6,7 @@ package context
 
 import (
 	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
 	macaron "gopkg.in/macaron.v1"
```
Comment (lunny) [inline modules/context/permission.go:11]: blank line.

File: models/access.go [hunk 3]
```diff
@@ -4,7 +4,11 @@
 
 package models
 
-import "fmt"
+import (
```
Comment (techknowlogick) [inline models/access.go:8]: Please add Gitea copyright to header of file

**Review by techknowlogick — APPROVED** [review]
One nit. Looks good otherwise.

File: models/access.go (hunk 3 above)
Comment (zeripath) [reply models/access.go:8]: Done

**Comment by adelowo** [comment]
Isn't `Trace` implementation supposed to be a no-op if `!log.IsTrace` rather than check if it's Trace level all the time ?

**Comment by zeripath** [comment]
Ah yes but constructing the trace call is not free - so if we can avoid constructing it at all that may make sense.

A lot of the calls there are less complex than I originally had them because I moved functionality into colorformat, so we could remove a few of the IsTrace calls but not all

**Review by adelowo — APPROVED** [review]

**Pull request merged**